### ISSUE TITLE
Pulumi precompiled Go programs still require go toolchain

### ISSUE THREAD

**Issue by t0yv0** [pulumi/pulumi#11167]
### What happened?

Pulumi requires a Go toolchain even when a precompiled binary is configured. This is done to "discover plugin requirements".

Scenario: deploy Pulumi program on an EC2 instance.

Because of the Go dependency:

- EC2 instance must have Go toolchain installed
- Initial pulumi preview takes a ton of time on EC2 because:
  - go is downloading large dependencies such as GRPC
  - go is compiling large dependencies

To reproduce:

    # clone https://gist.github.com/t0yv0/f56c90ccbedd1266a8108e12209c640e
    pulumi stack new repro
    pulumi stack select repro
    ./repro.sh

Expected: pulumi preview succeeds.

Actual: pulumi preview fails when trying to call `go` toolchain that
is swapped out with an fail-fast script. It will also fail if the `go`
toolchain is simply missing.

    error: failed to discover plugin requirements: determining go version: exit status 1

### Steps to reproduce

https://gist.github.com/t0yv0/f56c90ccbedd1266a8108e12209c640e

### Expected Behavior

Precompiled Go programs do not have Go toolchain as a requirement.

### Actual Behavior

Precompiled Go programs fail unless Go toolchain is available.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

**Comment by t0yv0** [comment]
True. This would need somehow compiling the plugin requirements into the binary itself.

**Comment by iwahbe** [comment]
We use the go toolchain to discover which plugins need to be installed instead of manually parsing `go.mod`. We should be able to make this process optional, deferring plugin discovery to runtime (when we get the plugin requests).

**Comment by abhinav** [comment]
Switching to kind/enhancement.

**Comment by dkrieger** [comment]
This causes unexpected errors when using a build system like bazel that doesn't require the host to have a go toolchain
